besu/src/main/java/org/hyperledger/besu/services/TransactionSimulationServiceImpl.java

```diff
@@ -51,50 +56,67 @@ public void init(final Blockchain blockchain, final TransactionSimulator transac
+    final var maybeBlockHeader =
+        blockchain.getBlockHeader(blockHash).or(() -> blockchain.getBlockHeaderSafe(blockHash));
 
-      final var maybeBlockHeader =
-          blockchain.getBlockHeader(blockHash).or(() -> blockchain.getBlockHeaderSafe(blockHash));
+    if (maybeBlockHeader.isEmpty()) {
+      return Optional.of(
+          new TransactionSimulationResult(
+              transaction,
+              TransactionProcessingResult.invalid(
+                  ValidationResult.invalid(TransactionInvalidReason.BLOCK_NOT_FOUND))));
+    }
 
-      if (maybeBlockHeader.isEmpty()) {
-        return Optional.of(
-            new TransactionSimulationResult(
-                transaction,
-                TransactionProcessingResult.invalid(
-                    ValidationResult.invalid(TransactionInvalidReason.BLOCK_NOT_FOUND))));
-      }
+    return transactionSimulator
+        .process(
+            callParameter,
+            isAllowExceedingBalance
+                ? transactionSimulatorAllowExceedingBalance()
+                : transactionSimulator(),
+            operationTracer,
+            maybeBlockHeader.get())
+        .map(res -> new TransactionSimulationResult(transaction, res.result()));
+  }
 
-      return transactionSimulator
-          .process(
-              callParameter,
-              isAllowExceedingBalance
-                  ? TransactionValidationParams.transactionSimulatorAllowExceedingBalance()
-                  : TransactionValidationParams.transactionSimulator(),
-              operationTracer,
-              maybeBlockHeader.get())
-          .map(res -> new TransactionSimulationResult(transaction, res.result()));
-    }
+  @Override
+  public Optional<TransactionSimulationResult> simulate(
+      final Transaction transaction,
+      final Optional<StateOverrideMap> maybeStateOverrides,
+      final ProcessableBlockHeader pendingBlockHeader,
+      final OperationTracer operationTracer,
+      final boolean isAllowExceedingBalance,
+      final boolean isAllowFutureNonce) {
+
+    final CallParameter callParameter = CallParameter.fromTransaction(transaction);
 
     return transactionSimulator
         .processOnPending(
             callParameter,
             maybeStateOverrides,
             isAllowExceedingBalance
-                ? TransactionValidationParams.transactionSimulatorAllowExceedingBalance()
-                : TransactionValidationParams.transactionSimulator(),
+                ? isAllowFutureNonce
+                    ? transactionSimulatorAllowExceedingBalanceAndFutureNonce()
+                    : transactionSimulatorAllowExceedingBalance()
+                : isAllowFutureNonce
+                    ? transactionSimulatorAllowFutureNonce()
+                    : transactionSimulator(),
             operationTracer,
-            transactionSimulator.simulatePendingBlockHeader())
+            (org.hyperledger.besu.ethereum.core.ProcessableBlockHeader) pendingBlockHeader)
```

Comment:
If we change the import in TransactionSimulator and BlockHashProcessor to org.hyperledger.besu.plugin.data.ProcessableBlockHeader, we do not need this casst

Reply:
I tried but found much more places to change and gave up